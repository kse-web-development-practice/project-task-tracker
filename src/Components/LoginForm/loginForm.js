import React, { useContext, useState } from 'react'
import styles from './loginForm.module.css'
import PropTypes from 'prop-types'
import { FormInput } from '../FormInput/formInput'
import { Button } from '../Button/button'
import { UserContext } from '../../user-context'
import { useNavigate } from 'react-router-dom'

export const LoginForm = ({ loginFunc }) => {
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState('')

  const userContext = useContext(UserContext)

  const navigate = useNavigate()

  async function handleSubmit(event) {
    event.preventDefault()
    if (!username || !password) {
      setError('Please fill in all fields.')
      return
    }
    const user = await loginFunc(username, password)
    if (!user) {
      setError('Incorrect username or password.')
      console.log('SHIT')
      return
    }
    userContext.setUser(user.username, user.token)
    navigate('/')
    setError('')
  }

  return (
    <form className={styles.loginForm} onSubmit={handleSubmit}>
      {error && <div className="error">{error}</div>}
      <FormInput
        labelText="Username:"
        type=""
        value={username}
        onChange={(e) => setUsername(e.target.value)}
      />
      <FormInput
        labelText="Password:"
        type="password"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <Button isMain={true} type="submit">
        Login
      </Button>
    </form>
  )
}

LoginForm.propTypes = {
  loginFunc: PropTypes.func
}
