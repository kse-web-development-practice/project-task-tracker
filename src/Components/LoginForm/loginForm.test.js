import React from 'react'
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'
import { LoginForm } from './loginForm'
import { useNavigate } from 'react-router-dom'
import { UserContext } from '../../user-context'

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn()
}))

describe('LoginForm Component', () => {
  const renderLoginForm = (loginFunc) =>
    render(
      <UserContext.Provider value={{ setUser: jest.fn() }}>
        <LoginForm loginFunc={loginFunc} />
      </UserContext.Provider>
    )

  test('default snapshot', () => {
    const { asFragment } = renderLoginForm()
    expect(asFragment()).toMatchSnapshot()
  })

  test('empty fields error snapshot', () => {
    const { asFragment } = renderLoginForm()
    act(() => {
      fireEvent.click(screen.getByText('Login'))
    })

    expect(asFragment()).toMatchSnapshot()
  })

  test('valid submission', async () => {
    const loginFunc = jest.fn().mockResolvedValue({ username: 'testUser', token: 'abc123' })
    const mockNavigate = jest.fn()
    useNavigate.mockReturnValue(mockNavigate)

    renderLoginForm(loginFunc)

    act(() => {
      fireEvent.change(screen.getByLabelText(/Username:/i), { target: { value: 'testUser' } })
      fireEvent.change(screen.getByLabelText(/Password:/i), { target: { value: 'password123' } })

      fireEvent.click(screen.getByText('Login'))
    })

    await waitFor(() => {
      expect(loginFunc).toHaveBeenCalledWith('testUser', 'password123')
      expect(mockNavigate).toHaveBeenCalledWith('/')
    })
  })

  test('incorrect credentials error snapshot', async () => {
    const loginFunc = jest.fn().mockResolvedValue(null)
    const { asFragment } = renderLoginForm(loginFunc)

    await act(async () => {
      fireEvent.change(screen.getByLabelText(/Username:/i), { target: { value: 'testUser' } })
      fireEvent.change(screen.getByLabelText(/Password:/i), { target: { value: 'wrongpassword' } })

      fireEvent.click(await screen.findByText('Login'))
    })

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
