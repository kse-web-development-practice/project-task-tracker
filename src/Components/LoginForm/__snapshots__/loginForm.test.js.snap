// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component default snapshot 1`] = `
<DocumentFragment>
  <form
    class="loginForm"
  >
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        type="password"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Login
    </button>
  </form>
</DocumentFragment>
`;

exports[`LoginForm Component empty fields error snapshot 1`] = `
<DocumentFragment>
  <form
    class="loginForm"
  >
    <div
      class="error"
    >
      Please fill in all fields.
    </div>
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        type="password"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Login
    </button>
  </form>
</DocumentFragment>
`;

exports[`LoginForm Component incorrect credentials error snapshot 1`] = `
<DocumentFragment>
  <form
    class="loginForm"
  >
    <div
      class="error"
    >
      Incorrect username or password.
    </div>
    <label
      class="formLabel"
    >
      Username:
      <input
        class="formInput"
        type=""
        value="testUser"
      />
    </label>
    <label
      class="formLabel"
    >
      Password:
      <input
        class="formInput"
        type="password"
        value="wrongpassword"
      />
    </label>
    <button
      class="button main"
    >
      Login
    </button>
  </form>
</DocumentFragment>
`;
